---
title: "CRISPR Lineage Tracing QC"
output: 
  html_document:
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
library(dplyr)
library(tidyr)
library(magrittr)
library(knitr)
library(kableExtra)
library(readr)
library(stringr)
library(ggplot2)
library(stringr)
library(fastqcr)
library(RColorBrewer)
library(traceQC)
library(circlize) # Added due to a function dependency, has to be out 
library(ComplexHeatmap) # Added due to a function dependency, has to be out 
``` 

## Data
Randomly sampled 5000 reads from Church's data.

## Plot construct
```{r construct}
ref <- system.file("extdata", "test_data","ref.txt",package = "traceQC")
plot_construct(ref)
```

## FastQC {.tabset}
```{r QC}
qc_dir <- tempdir()
fastqc(system.file("extdata", "test_data",package = "traceQC"),qc.dir=qc_dir)
```

### Per base quality
``` {r}
qc <- qc_read(file.path(qc_dir, "example_fastqc.zip"))
qc_plot(qc, "Per base sequence quality")
```

### Per sequence quality
``` {r}
qc_plot(qc, "Per sequence quality scores")
```

### Sequence length distribution
``` {r}
qc_plot(qc, "Per base N content")
```

## Alignment
``` {r align}
start_time <- Sys.time()

align_script <- system.file("py", "alignment.py", package="traceQC")
input_file <- system.file("extdata", "test_data", "example.fastq", package="traceQC")
ref_file <- system.file("extdata", "test_data", "ref.txt", package="traceQC")
output_file <- tempfile()
# system("python ./src/alignment.py --input ./test_data/example.fastq --reference ./test_data/ref.txt --output ./aligned.txt")
cmd <- paste("python", align_script, "--input", input_file, "--reference", ref_file, "--output", output_file)
system(cmd)
end_time <- Sys.time()
paste("Aligning 5000 reads in", round(end_time - start_time,2))
```

## Output file
``` {r out}
aligned_barcode <- read_tsv(output_file) %>%
  mutate(id=1:n())
print(nrow(aligned_barcode))
kable(head(aligned_barcode) %>% select(id,score,spacer_seq)) %>% # count is not available
  kable_styling(bootstrap_options = "striped", full_width = F)
```

## Score distribution
``` {r score}
ggplot(aligned_barcode) + 
  geom_histogram(aes(x=score,y=..density..),binwidth=5) +
  ylab("percentage") +
  theme_classic()
```

## Lorenz curve
``` {r lorenz}
lorenz_curve(aligned_barcode)
```

## Sequence to character
``` {r s2c}
start_time <- Sys.time()
out_path <- tempfile()
mutation_df <- seq_to_character(aligned_barcode, out_path) %>%
  mutate(id=1:n())
end_time <- Sys.time()
paste("It takes",end_time-start_time)
kable(head(mutation_df) %>% select(id,count,type,start,length))
```

## Number of mutations per barcode
``` {r}
num_mutation_histogram(mutation_df)
```

## Mutation types
``` {r,message=FALSE}
mutation_type(mutation_df) 
```

## Mutation Hotspot {.tabset}

### Deletions
``` {r}
deletions <- mutation_df %>%
  filter(type=="deletion") %>%
  group_by(start,length) %>%
  summarise(count=sum(count)) %>%
  mutate(end=start+length) %>%
  ungroup %>%
  mutate(id=1:n()) %>%
  mutate(log10_count=log10(count))
circular_chordgram(df=deletions,title="deletions",ref)
```

### Insertions
``` {r}
insertions <- mutation_df %>%
  filter(type=="insertion") %>%
  group_by(start,length) %>%
  summarise(count=sum(count)) %>%
  mutate(end=start+length) %>%
  ungroup %>%
  mutate(id=1:n()) %>%
  mutate(log10_count=log10(count))

circular_chordgram(df=insertions,title="insertions",ref)
```
### Mutations
``` {r}
mutations <- filter(mutation_df,type=="mutation") %>%
  group_by(start,length,mutate_to) %>%
  summarise(count=sum(count)) %>%
  ungroup

circular_histogram(mutations,ref)

```
