---
title: "CRISPR Lineage Tracing QC"
output: 
  html_document:
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
library(dplyr)
library(tidyr)
library(magrittr)
library(knitr)
library(kableExtra)
library(readr)
library(stringr)
library(ggplot2)
library(stringr)
library(fastqcr)
library(RColorBrewer)
library(parallel)
library(TraceQC)
library(reticulate)
library(circlize) # Added due to a function dependency, has to be out 
library(ComplexHeatmap) # Added due to a function dependency, has to be out 
``` 

## Data
Randomly sampled 5000 reads from Church's data.

## FastQC {.tabset}
```{r QC}
qc_dir <- tempdir()
fastqc(system.file("extdata", "test_data",package = "TraceQC"),qc.dir=qc_dir)
```

### Per base quality
``` {r}
qc <- qc_read(file.path(qc_dir, "example_fastqc.zip"))
qc_plot(qc, "Per base sequence quality")
```

### Per sequence quality
``` {r}
qc_plot(qc, "Per sequence quality scores")
```

### Sequence length distribution
``` {r}
qc_plot(qc, "Per base N content")
```

## Alignment
``` {r align}
start_time <- Sys.time()

input_file <- system.file("extdata", "test_data", "example.fastq", package="TraceQC")
ref_file <- system.file("extdata", "test_data", "ref.txt", package="TraceQC")
output_file <- tempfile()
sequence_alignment(input_file,ref_file,output_file=output_file,
                      match=2,mismatch=-2,gapopen=-6,gapextension=-0.1)
end_time <- Sys.time()
paste("Aligning 5000 reads in", round(end_time - start_time,2))
```

## Output file
``` {r out}
TraceQC_input <- create_input_object(output_file,ref_file)
```

## Plot construct
```{r construct}
plot_construct(TraceQC_input)
```

## Score distribution
``` {r score}
plot_score_distribution(TraceQC_input)
```

## Lorenz curve
``` {r lorenz}
lorenz_curve(TraceQC_input)
```

## Sequence to character
``` {r s2c}
start_time <- Sys.time()
out_path <- tempfile()
mutation_df <- seq_to_character(TraceQC_input, out_path)
end_time <- Sys.time()
paste("It takes",end_time-start_time)
```

## Number of mutations per barcode
``` {r}
num_mutation_histogram(mutation_df)
```

## Mutation types
``` {r,message=FALSE}
mutation_type(mutation_df)
```

## Mutation Hotspot {.tabset}

### Deletions
``` {r}
plot_deletion_hotspot(mutation_df,TraceQC_input)
```

### Insertions
``` {r}
plot_insertion_hotspot(mutation_df,TraceQC_input)
```
### Mutations
``` {r}
plot_point_mutation_hotspot(mutation_df,TraceQC_input)
```

## Build phylo tree
``` {r}
df <- filter(mutation_df,count>5,type!="unmutated")
tree_input <- build_character_table(df)

library(phangorn)
library(ggtree)
data <- phyDat(data=tree_input$character_table,type="USER",levels=c(0,1))
dm <- dist.hamming(data)
treeUPGMA <- upgma(dm)
treePars <- optim.parsimony(treeUPGMA, data)

ggtree(treePars) +
  geom_tiplab()
```

